Add external calling functionality

bwasti opened the pull request:
This PR adds the ability to register external functions directly into the IR.

An example, matmul, is used to show how this registration is done.

This will bootstrap operators that are currently (or always) hard to generate.  It also enables composability with other specialized compilers (such as FBGEMM).

zheng-xq committed:
refcount fixing self-assignment

zheng-xq committed:
Make LOG(FATAL) nonreturn
Enable Werror

zheng-xq committed:
Adding statement conversion for SplitWithTail

zheng-xq committed:
Add a reference tests for Split

bertmaher committed:
clang-format

zheng-xq committed:
A functinoal reference chck for schedule tests.

zheng-xq committed:
clang-format

Commit:
Add support for Float immediates.

bertmaher committed:
Get absolute path for ASMJIT_DIR (#24)

bertmaher committed:
Silence deprecation warnings from LLVM

bertmaher committed:
Include legacy PassManager for debug printing

bertmaher committed:
Set code model to medium to avoid indirect jumps in generated asm

bertmaher committed:
Fix argument type of input float buffers

Commit:
Add support for Casts in LLVM codegen.

Commit:
Enable export of compile_commands.json.

zheng-xq committed:
Add a complete tensor+lower+llvm test

zheng-xq committed:
Enable the failing test

bertmaher committed:
Floating point arithmetic

bertmaher committed:
Test fp32 mul using compute expr

bertmaher committed:
Broadcast add test using compute expr

Commit:
Update to LLVM 9

Commit:
Implementation of Broadcast for LLVM.

Commit:
Cleanup use of ConstantInt API.

zheng-xq committed:
Add Buffer operator() overload, and some other minor features

zheng-xq committed:
fix accidental experimental changes

zheng-xq committed:
Change the Compute interface to bring the dim sizes and names together

zheng-xq committed:
clang-format

zheng-xq committed:
refactor Buffer into its own files

zheng-xq committed:
Add a PaddedBuffer test util

Commit:
Add support for vector casts in LLVM CodeGen

Commit:
Implement masked loads and stores.

Commit:
Implement vector masked loads and stores.

zheng-xq committed:
Improve the user interface for SimpleIREvaluator

Commit:
Add a test for Block codegen.

Commit:
Fix gtest include path

bertmaher committed:
clang-format

resistor committed:
Add expressions and support for Max and Min. (#5)

ZolotukhinM committed:
Rename compiler to tensorexpr and move files around to be more similar to other pytorch parts. (#6)

Summary:

1. Move compiler to tensorexpr folder
2. Move files from src and include to the same folder (and remove src and include folders)
3. Rename .cc to .cpp

ZolotukhinM committed:
Add missing include <math.h> (#7)

zheng-xq committed:
Change isnan to std::isnan. It breaks my clang builds. (#8)

zheng-xq committed:
Change the SimpleIREvaluator frontend (#9)

Add RefHandle for subclass

resistor committed:
Make LLVM dependency optional. (#10)

bertmaher committed:
[wip] Basic fuser pass to select texpr subgraphs

bertmaher committed:
Revert "[wip] Basic fuser pass to select texpr subgraphs"

This reverts commit a9d9919b0570220772bafcb3667c9bee6b90d051.

resistor committed:
Add a test for aten::_cast_Float lowering. (#12)

Commit:
Revert changes to the main pytorch CMakeLists.txt (for now).

bertmaher committed:
Hook tensorexp up to the main build, and switch to c10 logging

resistor committed:
More ATen op tests. (#16)

bertmaher committed:
Fix some missing returns

ZolotukhinM committed:
Include tests back to the 'all' target. (#14)

resistor committed:
Even more ATen op tests. (#18)

resistor committed:
Test for relu ATen op. (#19)

zheng-xq committed:
Add intrinsics function support. (#20)

zheng-xq committed:
Remove fmax/fmin, as they are already covered by the Max/Min operators (#21)

zheng-xq committed:
refactor CallNode and BaseCallNode, so we can have a common concrete base class for visitors. (#22)

This is the first step to add other call types.

zheng-xq committed:
Add FunctionCall to use existing tensors (#23)

zheng-xq committed:
Add the ability to use an existing tensor expression in other compute functions. (#24)

Commit:
fixing broken compilation on mac/clang

Commit:
adding IRnode for Compare-Select Ops and their LLVM Codegen

bertmaher committed:
Tensor expr fuser pass for extremely simple expressions

resistor committed:
Fix Werror. (#26)

resistor committed:
Add tests for some transcendental ops. (#27)

zheng-xq committed:
Add Allocate and Free support. (#29)

Add Eval and test basic alloc support.
Add Lowering support for buffer allocation for intermediate tensors.

bertmaher committed:
Make fusion work for arbitrary buffer/tensor combinations of inputs (#30)

Commit:
fix Let02 test

bertmaher committed:
Access inputs and intermediates uniformly through Tensors (#31)

bertmaher committed:
Make tx fuser work with arbitrary ranks

Commit:
adding LLVM Codegen for Let

bertmaher committed:
[fuser] Broadcast args

zheng-xq committed:
Adding ComputeInline support. (#35)

zheng-xq committed:
Fix broken tests (#36)

bertmaher committed:
Improve naming of arg broadcasting function

Commit:
modifying CMakeLists.txt to enable ninja test && minor update for LLVM Codegen for Let (handling XQ's comment)

bertmaher committed:
Test cases for tensorexpr fusion (#37)

Commit:
CompareSelct Op: Addressing XQ and Owen's comments

resistor committed:
Sketch sufficient support for constants to get constant alpha working. (#40)

* Refactor to use a switch statement over Node kinds.

* Sketch sufficient support for constants to get constant alpha working.

bertmaher committed:
Fix indices when inlining non-leaf calls (#39)

zheng-xq committed:
Fixing the inline ordering issue (#43)

Solve more problems with the inliner

resistor committed:
Avoid creating redundant and/or improperly ordered Constant's in fused subgraphs. (#42)

bertmaher committed:
Move fuser-styled tests to schedule_test (#44)

resistor committed:
Add aten::sub to the new fuser. (#46)

zheng-xq committed:
Refactor CodeGen from SimpleIREval (#47)

bertmaher committed:
Inline all the things (#45)

Commit:
clang-format for atent_test.cpp

resistor committed:
Eliminate a ton of warnings for my own sanity. (#48)

Commit:
aten tests for eq, ge, gt, le, lt

resistor committed:
Add support for type promotion/demotion. (#50)

resistor committed:
Flesh out new fuser coverage to several more ops. (#51)

Commit:
support for aten ops: eq

Commit:
support for more  aten ops: ge, gt, le, lt, ne

zheng-xq committed:
Adding the first basic CudaCodeGen. (#52)

bertmaher committed:
Minimal CMake change to link LLVM to libtorch

bertmaher committed:
Fix issues causing assertion failures in llvm debug builds

bertmaher committed:
Fatal on unimplement llvm codegen ops (Allocate, etc.)

bertmaher committed:
Optionally compile tx fuser kernels with llvm

bertmaher committed:
Test for 2D broadcasted with large dims to show vectorization

bertmaher committed:
Refactor LLVMCodeGen to compile kernel in constructor

resistor committed:
Updated isSupported for increased op coverage. (#54)

ZolotukhinM committed:
Cmake integration to PT codebase (#28)

With this change our code blends with the usual PyTorch code and is built the usual way. I added a cmake option to specify where to look for LLVM, if it's not specified, LLVM is not used.

An example of invocation (from the root of pytorch repo):

```
USE_LLVM=/path/to/llvm9/install  python setup.py develop
```

This command will build libtorch.{a,so} and other libraries, and tensorexpr code will be a part of it.

The tests will be built in build/bin/test_tensorexpr (I've ported only one test so far). So, invocation of the tests will be:

```
build/bin/test_tensorexpr
```

ZolotukhinM committed:
Remove old padded_buffer.{cpp,h}. (#56)

resistor committed:
Add support for code generation of Log10 intrinsics with LLVM. (#57)

ZolotukhinM committed:
Remove tests/test_utils.h: inline what's still used and nuke what's unused. (#58)

ZolotukhinM committed:
Move Fuser tests (tests/tests.py) to test/test_tensorexpr.py. (#59)

Commit:
Remove old CMakeLists and README.txt

resistor committed:
Add support for vectorized and unmasked loads and stores with LLVM. (#62)

resistor committed:
Enable CodeGen-level optimizations in LLVM. (#63)

zheng-xq committed:
Add Bind/GPUBlock/GPUThread support. (#64)

bertmaher committed:
Bind/run interface to CodeGen (#60)

* Bind/run interface to CodeGen

* Make LLVMCodeGen implement CodeGen interface

* Allow bind/run to be unimplemented for the moment (CUDA)

* Cache compilation result

* Two nasty bugs: forgot virtual dtor, forgot to clear bindings after run()

bertmaher committed:
Fix ambiguity in CreateExtractElementCall (0ull can be a Value*, I guess?) (#65)

bertmaher committed:
Allow constants as lhs/rhs args (not just alpha) (#66)

bertmaher committed:
Use correct tensor type for fuser output (#67)

bertmaher committed:
clang-format

Commit:
Rename 'compiler' namespace to 'tensorexpr'.

bertmaher committed:
Include all built llvm targets (#68)

resistor committed:
Switch back to linking only the native LLVM target. (#69)

bertmaher committed:
Virtual dtors for IRVisitor/IRMutator (#70)

zheng-xq committed:
Add semicolon to make nvcc compile (#71)

zheng-xq committed:
Enable NVRTC for the GPU backend. (#74)

resistor committed:
Fix non-CUDA testing. (#75)

protonu committed:
Getting fused (a)Sin(h), (a)Cos(h),(a) Tan(h), abs working with the interpreter (#73)

* Getting fused (a)Sin(h), (a)Cos(h),(a) Tan(h), abs working with the interpreter

* take the interpreter path only when ENABLE_LLVM is not set

zheng-xq committed:
remove the leak tests, as we will get rid of refcounting (#76)

bertmaher committed:
Implement aten::min, max, and clamp (#72)

* Implement aten::min, max, and clamp

* Propagate NaNs like std::max/min

* Change NaN propagation in interpreter too

bertmaher committed:
clang-format tensorexpr/tests.h (#77)

zheng-xq committed:
Refactor UniqueNameManager into its own files. (#79)

zheng-xq committed:
refactor cuda_codegen (#80)

zheng-xq committed:
simplify nvrtc major, minor versions (#81)

bertmaher committed:
Allow CodeGen to take Var args (interpreter support only) (#78)

* Test demonstrating dynamic shape

* Allow binding of Vars to args in interpreter

* Pass BufferArgs to LLVMCodeGen

* clang-format-diff

bertmaher committed:
[LLVMCodeGen] Refactor kernel constructor to be less sprawling (#82)

* Member TM to TM_ in LLVMCodeGen

* [LLVMCodeGen] Add helper for getContext

* [LLVMCodeGen] Refactor type support

* [LLVMCodeGen] Refactor kernel emission

protonu committed:
(TE Interpreter)Support for floor, ceil, trunc, remainder, sqrt and improving tests  (#83)

* Getting fused (a)Sin(h), (a)Cos(h),(a) Tan(h), abs working with the interpreter
* take the interpreter path only when ENABLE_LLVM is not set
* cleaning up the tests for the new aten ops
* (TE Interpret)adding support for floor, ceil, trunc, remainder and improving tests

zheng-xq committed:
Add Cond and Mod to SimpleIREval (#84)

bertmaher committed:
[LLVMCodeGen] Support dynamic shapes by binding Var args (#86)

* [LLVMCodeGen] Support dynamic shapes by binding Var args

* Test llvm dynamic shape codegen using Tensor

zheng-xq committed:
Add SplitWithMask core support. (#87)

zheng-xq committed:
Add Cuda tests for SplitWithMask (#88)

zheng-xq committed:
Disable DEBUG_PRINT (#89)

zheng-xq committed:
Remove some debug prints (#90)

resistor committed:
Fix the no-CUDA build. (#92)

resistor committed:
Add support for multiple outputs from the fused subgraph. (#91)

zheng-xq committed:
Remove RefCounting (#93)

zheng-xq committed:
Add some comments for KernelScope. Address comments. (#94)

zheng-xq committed:
Completely remove refcount.h (#95)

zheng-xq committed:
fix the fuser pass (#97)

zheng-xq committed:
Rename Kernel to KernelArena (#98)

resistor committed:
Add support for fusion through ConstantChunk ops. (#96)

resistor committed:
Fix implicit noexcept deduction warning. (#99)

bertmaher committed:
Make llvm tests conditional on USE_LLVM (#100)

* Make llvm tests conditional on USE_LLVM

* Use the right macro and add to gtest harness

* clang-format

resistor committed:
Refactor ComputeNode into ComputeValue, to be able to handle arbitrary (#101)

multi-output operators.

bertmaher committed:
Improve Stmt pretty printing from TensorExprFuser (#102)

resistor committed:
Add support for IfThenElse (#103)

zheng-xq committed:
Add end-to-end support and a PyTorch fuser example on CudaCodeGen (#104)

zheng-xq committed:
fix rebase errors (#105)

protonu committed:
fixes to build on system without LLVM and CUDA (#107)

* fixes to build on system without LLVM and CUDA

* minor edit: fixes to build on system without LLVM and CUDA

resistor committed:
Add support for aten::cat to the new fuser. (#106)

resistor committed:
Bail out of fusion if we don't have a complete tensor type (for now). (#108)

bertmaher committed:
Standardize codegen call() interface and remove bind/run (#109)

* Standardize codegen call() interface and remove bind/run

* revert undef USE_CUDA

bertmaher committed:
Clean up sketchy handling of scalar args in llvm codegen (#110)

bertmaher committed:
Test 2D dynamic shapes (#112)

bertmaher committed:
clang-format (#113)

resistor committed:
Add LLVM codegen for a lot of transcendental ops. (#115)

resistor committed:
Fix bug with binary math intrinsics. (#116)

bertmaher committed:
Use CUDA for 3-arg test (#117)

zheng-xq committed:
Refactor CudaCodeGen into generic registration, so we can have both the Cuda and non-Cuda builds. (#118)

Commit:
Add instructions on how to rebase on master.

bertmaher committed:
Dynamic shape support in CUDA codegen (#120)

* Dynamic shape support in CUDA codegen

* free cuda memory

zheng-xq committed:
Disable GPU fuser. Revive the Cuda tests (#121)

zheng-xq committed:
Add ExecutionCounter to detect whether the underlying code is executed. (#122)

zheng-xq committed:
Adding GPU index flatting to support arbitrary elementwise and broadcasting support. (#126)

Krovatkin committed:
fix a bug kLog to Intrin::log (#124)

bertmaher committed:
Allow scalar variables as inputs (#125)

bertmaher committed:
clang-format (#127)

bertmaher committed:
Format python tests with `black` (#128)

resistor committed:
Add support for fusion in nested blocks. (#129)

resistor committed:
Teach the LLVM JIT to use dlsym to resolve symbols. (#130)

bertmaher committed:
Factor out kernel codegen from tx fusion pass (#131)

Commit:
Use standard JIT logging in TX fuser.

ZolotukhinM committed:
Move memory management classes (KernelArena, KernelScope, KernelScopedObject) to a separate file. (#132)

protonu committed:
(IR Interpreter) Adding more Operators: Erfc, Exmp1, frac, lgamma, neg, sigmoid, reciprocal, neg, relu (#133)

bertmaher committed:
Add erfc to llvm codegen (#134)

bertmaher committed:
Squash some warnings (#135)

protonu committed:
(IR interpreter) addcmul (#137)

* (IR interpreter) addcmul

zheng-xq committed:
Remove IRNode. CodeGen accepts only Stmt. Add ExprEval utility wrapper. (#138)

zheng-xq committed:
Add the benchmark from NNC (#141)

resistor committed:
Fix verifier errors in LLVM codegen when conditional loads feed directly into concats. (#143)

resistor committed:
Strength reduction peephole for pow(). (#144)

resistor committed:
Fix incorrect pow(x, 0) case. (#145)

bertmaher committed:
Use `const Value*` where possible (#146)

zheng-xq committed:
Make Broadcast work (#147)

$ python benchmarks/tensorexpr/benchmark.py broadcast_3args --device gpu --mode fwd --jit_mode trace

zheng-xq committed:
Fixed CudaCodeGen output streams. Switch to __ldg by default (#148)

zheng-xq committed:
Add ElementWise support (#150)

resistor committed:
Fix an assertion failure when merging constants into aten::cat fusions. (#151)

protonu committed:
adding LLVM support ops: sigmoid, relu, neg, addcmul, reciprocal, lgamma, expm1 (#149)

* adding LLVM support for a few ops

lly-zero-one committed:
Add more operator support and tests (#140)

* Add more operator support and tests

rm log

add more cuda tests

clean up debug

relu is already added

fix the frac/relu support

* rm the extra relu

* redundant op

* rm frac

bertmaher committed:
Fix accidental assignment in condition (#153)

zheng-xq committed:
Add elementwise benchmarks and comparisons. (#155)

ZolotukhinM committed:
Backport some of the fixes from the master PR. (#157)

zheng-xq committed:
Adding broadcasting benchmarks (#158)

zheng-xq committed:
Fix the missing aten::pow support (#160)

zheng-xq committed:
Fix the missing aten::pow support (#161)

Add "f" in floating point literal

zheng-xq committed:
Fixing the failing test (#164)

resistor committed:
Add NNC support for aten::slice and aten::unsqueeze. (#159)

bertmaher committed:
Get strides working (#163)

* Get strides working

* Handle (only) constant strides

zheng-xq committed:
Change the default block size from 1024 to 512. (#165)

$ (PYTHONPATH=$PWD python benchmarks/tensorexpr/benchmark.py element broadcast --device gpu --mode fwd --jit_mode trace) |& tee /tmp/01.txt
pt: element_split_mul_fwd_cuda_134217728: 3281.45 us, SOL 818.04 GB/s, algorithmic 818.04 GB/s
pt: element_split_add_fwd_cuda_134217728: 3282.54 us, SOL 817.77 GB/s, algorithmic 817.77 GB/s
pt: element_split_sub_fwd_cuda_134217728: 3282.69 us, SOL 817.73 GB/s, algorithmic 817.73 GB/s
pt: element_split_div_fwd_cuda_134217728: 3308.66 us, SOL 811.31 GB/s, algorithmic 811.31 GB/s
pt: element_split_pow_fwd_cuda_134217728: 3516.52 us, SOL 763.35 GB/s, algorithmic 763.35 GB/s
pt: element_split_max_fwd_cuda_134217728: 3282.74 us, SOL 817.72 GB/s, algorithmic 817.72 GB/s
pt: element_split_min_fwd_cuda_134217728: 3282.73 us, SOL 817.72 GB/s, algorithmic 817.72 GB/s
pt: element_shared_mul_fwd_cuda_134217728: 1356.76 us, SOL 791.40 GB/s, algorithmic 791.40 GB/s
pt: element_shared_add_fwd_cuda_134217728: 1356.73 us, SOL 791.42 GB/s, algorithmic 791.42 GB/s
pt: element_shared_sub_fwd_cuda_134217728: 1356.69 us, SOL 791.44 GB/s, algorithmic 791.44 GB/s
pt: element_shared_div_fwd_cuda_134217728: 1375.88 us, SOL 780.40 GB/s, algorithmic 780.40 GB/s
pt: element_shared_pow_fwd_cuda_134217728: 2296.72 us, SOL 467.51 GB/s, algorithmic 467.51 GB/s
pt: element_shared_max_fwd_cuda_134217728: 1356.51 us, SOL 791.55 GB/s, algorithmic 791.55 GB/s
pt: element_shared_min_fwd_cuda_134217728: 1356.51 us, SOL 791.55 GB/s, algorithmic 791.55 GB/s
pt: element_split_exp_fwd_cuda_134217728: 3289.87 us, SOL 815.95 GB/s, algorithmic 815.95 GB/s
pt: element_split_sin_fwd_cuda_134217728: 3478.14 us, SOL 771.78 GB/s, algorithmic 771.78 GB/s
pt: element_split_cos_fwd_cuda_134217728: 3479.85 us, SOL 771.40 GB/s, algorithmic 771.40 GB/s
pt: element_shared_exp_fwd_cuda_134217728: 1377.77 us, SOL 779.33 GB/s, algorithmic 779.33 GB/s
pt: element_shared_sin_fwd_cuda_134217728: 1826.81 us, SOL 587.77 GB/s, algorithmic 587.77 GB/s
pt: element_shared_cos_fwd_cuda_134217728: 1840.52 us, SOL 583.39 GB/s, algorithmic 583.39 GB/s
pt: broadcast_split_mul_fwd_cuda_256_128_512: 111.39 us, SOL 602.47 GB/s, algorithmic 602.47 GB/s
pt: broadcast_split_add_fwd_cuda_256_128_512: 111.63 us, SOL 601.18 GB/s, algorithmic 601.18 GB/s
pt: broadcast_split_sub_fwd_cuda_256_128_512: 111.69 us, SOL 600.83 GB/s, algorithmic 600.83 GB/s
pt: broadcast_split_div_fwd_cuda_256_128_512: 166.97 us, SOL 401.91 GB/s, algorithmic 401.91 GB/s
pt: broadcast_split_pow_fwd_cuda_256_128_512: 310.92 us, SOL 215.84 GB/s, algorithmic 215.84 GB/s
pt: broadcast_split_max_fwd_cuda_256_128_512: 112.62 us, SOL 595.91 GB/s, algorithmic 595.91 GB/s
pt: broadcast_split_min_fwd_cuda_256_128_512: 112.58 us, SOL 596.09 GB/s, algorithmic 596.09 GB/s
pt: broadcast_shared_mul_fwd_cuda_256_128_512: 109.32 us, SOL 613.85 GB/s, algorithmic 613.85 GB/s
pt: broadcast_shared_add_fwd_cuda_256_128_512: 109.37 us, SOL 613.60 GB/s, algorithmic 613.60 GB/s
pt: broadcast_shared_sub_fwd_cuda_256_128_512: 109.36 us, SOL 613.63 GB/s, algorithmic 613.63 GB/s
pt: broadcast_shared_div_fwd_cuda_256_128_512: 162.10 us, SOL 414.00 GB/s, algorithmic 414.00 GB/s
pt: broadcast_shared_pow_fwd_cuda_256_128_512: 303.95 us, SOL 220.79 GB/s, algorithmic 220.79 GB/s
pt: broadcast_shared_max_fwd_cuda_256_128_512: 110.34 us, SOL 608.20 GB/s, algorithmic 608.20 GB/s
pt: broadcast_shared_min_fwd_cuda_256_128_512: 110.20 us, SOL 608.95 GB/s, algorithmic 608.95 GB/s
pt: broadcast_split_exp_fwd_cuda_256_128_512: 146.12 us, SOL 459.27 GB/s, algorithmic 459.27 GB/s
pt: broadcast_split_sin_fwd_cuda_256_128_512: 224.09 us, SOL 299.47 GB/s, algorithmic 299.47 GB/s
pt: broadcast_split_cos_fwd_cuda_256_128_512: 225.54 us, SOL 297.54 GB/s, algorithmic 297.54 GB/s
pt: broadcast_shared_exp_fwd_cuda_256_128_512: 141.29 us, SOL 474.97 GB/s, algorithmic 474.97 GB/s
pt: broadcast_shared_sin_fwd_cuda_256_128_512: 220.59 us, SOL 304.22 GB/s, algorithmic 304.22 GB/s
pt: broadcast_shared_cos_fwd_cuda_256_128_512: 221.44 us, SOL 303.06 GB/s, algorithmic 303.06 GB/s

ZolotukhinM committed:
Check that dtype is float before calling std::isnan. (#167)

ZolotukhinM committed:
Check that dtype is float before calling std::isnan. (#168)

ZolotukhinM committed:
Remove asmjit backend. (#169)

ZolotukhinM committed:
Cleanup fuser pass a little. (#170)

lly-zero-one committed:
Add the Binary/unary op and also tests (#154)

* Add the atan2 op

resolve conflict

remove the atan2 from the unary op

* Add the threshold op

* support addcmul for cuda

* add more binary ops

zheng-xq committed:
Adding options to set the cuda loop levels, block count and block size. (#172)

protonu committed:
fixing failures in LLVM codegen for Compare Select Ops (#173)

* fixing failures in LLVM codegen for Compare Select Ops and changing LLVM codegen to use ordered comparision

zheng-xq committed:
Add LetStmt support. (#174)

bertmaher committed:
Pass Graph to TensorExprKernel constructor (#177)

bertmaher committed:
Broadcast based on input shapes (#178)

zheng-xq committed:
Add PrioritizeLoad to CudaCodeGen. (#179)

protonu committed:
[WIP] Adding 4-Op CompareSelect (#175)

* [WIP] Adding 4-Op CompareSelect

ZolotukhinM committed:
Remove class FunctionNode. (#180)

resistor committed:
Add support for pow() in the LLVM backend. (#182)

resistor committed:
Add support for None operands to aten::clamp in the TE fuser. (#181)

resistor committed:
Add guard elimination support for aten::unsqueeze. (#33371) (#184)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/33371

Differential Revision: D19920041

Pulled By: resistor

fbshipit-source-id: 906af47676dba014c31eef069a4753207f2efc60

protonu committed:
fix for test testATengeInt (#185)

zheng-xq committed:
Adding Cuda Random support in TE. (#183)

* Adding Cuda Random support in TE.

Test Plan:
  * Run the test for 1000 times, and no flaky failures.
  * (build/bin/test_tensorexpr --gtest_filter="*Rand*" --gtest_repeat=1000)

ZolotukhinM committed:
Remove TensorNode and TensorOperationNode classes and remove some wrapper accessors to make the code more explicit. (#186)

* Remove wrapper function accessors from TensorNode: instead access function_'s members directly through function().

* Remove TensorNode class.

* Remove TensorOperationNode class.

Commit:
Add external calling functionality

zheng-xq committed:
Fix the broken Cuda build (#188)

bertmaher committed:
initial impl of symbolic shapes (#176)

* formatted guard elimination

* initial impl of symbolic shapes

bertmaher reviewed (commented):
This is super cool.  (I can do a real review on Monday but this is exciting :) )

opti-mix commented:
@bwasti Very cool!

A small proposal: Could we avoid using a global registry? It would be nice to have some kind of TensorExprEngineContext or something like that, which would contain all such registries and other settings specific to a given instance of the TensorExprEngine. The idea is to make it possible to invoke TensorExprEngine from different places with different contexts if required.

zheng-xq committed:
Add rand_like support, and Python tests (#189)

bertmaher committed:
Support dynamic shapes in texpr fuser (#190)

ZolotukhinM committed:
Use BaseExprNode* in IR classes directly rather than through Expr. (#191)

* Remove BaseStmtNode class.

* Use `const BaseExprNode*` instead of Expr in classes from ir.h

* Rename Expr->ExprHandler, Var->VarHandler, BaseExprNode->Expr, Variable->Var.

* Fixup CUDA build.

* Rename {Expr,Var}Handler to {Expr,Var}Handle.

* Fixup after rebase.

ZolotukhinM committed:
Backport a clang-tidy fix: replace BINARY_ACCEPT with IRPrinter::visitBinaryOp. (#192)

* Backport a clang-tidy fix: replace BINARY_ACCEPT with IRPrinter::visitBinaryOp.

* Make visitBinaryOp a local function rather than a method of IRPrinter.

ZolotukhinM committed:
Backport some changes from master. (#193)

ZolotukhinM committed:
Reenable the existing fuser by default and disable it only in our tests. (#194)

All `test_*` functions are now moved into a test-class (with no changes to them).

zheng-xq committed:
Add rand benchmark. (#196)

* Add rand benchmark.
* Add an option to disable texpr fuser.

lly-zero-one committed:
Add the cast_float, sigmoid_backward, tanh_backward and also fix the remainder (#198)

* Add the cast_float, backward ops and also fix the remainder

* fix the conflict

* change expr to exprhandle

* formatting

* fix the linter

bertmaher committed:
Fix some ir printer bugs (#201)

* Fix some ir printer bugs

* also true_stmt

bertmaher committed:
Enable axis splitting and GPU grid binding with variable shapes (#142)

* Enable axis splitting and GPU grid binding with variable shapes

* farwell ExprStmt, we hardly knew ye

ZolotukhinM committed:
Add a doc about end-to-end tensor expressions workflow. (#195)

* Add workflow.md.

* Remove the suggestions from the doc.

* Add language reference.

* Add language reference.

* Address some of the comments.

protonu committed:
Adding bitwise integer ops: &,^,<<, >> (#202)

* Adding bitwise integer ops: &,^,<<, >>

ZolotukhinM committed:
Replace ExprHandle with Expr* in Function, Tensor, and Buffer. (#200)

lly-zero-one committed:
Add the type_as support (#199)

* Add the cast_float, backward ops and also fix the remainder

fix the conflict

change expr to exprhandle

formatting

fix the linter

add the type_as support

* fix the threshold failure

protonu committed:
Aten op: where (#197)

* Aten op: where
This require a helper function which does promote types for the condition expression.

protonu committed:
LLVM codgen for fmod, remainder (#206)

* LLVM codgen for fmod, remainder

protonu committed:
fix testATengeInt (#208)

* fix testATengeInt

bertmaher committed:
Revert "initial impl of symbolic shapes (#176)"

This reverts commit 5bf52fa87879d70634b6dd4ad5b508a1f325f432.

ZolotukhinM committed:
Make functions actually support multiple outputs. (#204)

ZolotukhinM committed:
Move Stmt classes to a separate file. (#209)

The moved code wasnt changed.

nickgg committed:
Add support for more dtypes (#205)

ZolotukhinM committed:
Add indentation to IRPrinter's output. (#211)

bertmaher committed:
Log unsupported data types (#212)

resistor committed:
Fix warnings and errors when building with clang. (#215)

Krovatkin committed:
better logging in tensorexpr_fusion (#213)

nickgg committed:
use c10 type promotion rules (#214)

nickgg committed:
Fix typecasts in type_test (#216)
